torch/_library/autograd.py

```diff
@@ -31,8 +34,136 @@ def backward(ctx, *grads):
         },
     )
 
+    schema = opdef._opoverload._schema
+    if any(
+        utils.is_tensorlist_like_type(a.type)
+        for a in (*schema.arguments, *schema.returns)
+    ):
+        Generated = supports_tensorlist(Generated)
+
     def autograd_impl(*args):
         result = Generated.apply(*args)  # type: ignore[attr-defined]
         return result
 
     return autograd_impl
+
+
+def supports_tensorlist(cls: Any) -> Any:
+    """Allows a given autograd.Function class to support List[Tensor] inputs/outputs.
+
+    Regular autograd.Function has a constraint that it only directly supports autograd for
+    Tensors. Applying @supports_tensorlist enables an autograd.Function to support
+    autograd for List[Tensor] inputs and outputs.
+    """
+    # NB: All calls to the autograd.Function.apply shares these variables
+    # We assume that only one call to .apply happens at a time. This means that
+    # you cannot call the autograd.Function recursively (e.g. from its own forward).
+    input_spec: Optional[spec_t] = None
+    output_spec: Optional[spec_t] = None
+    result_is_tuple = None
```

Comment:
I think you can simplify most of these by:
- Returning specs/is_tuple as output to the forward. Any non-Tensor will just be ignored.
- Use a single nonlocal boolean flag to detect re-entrancy and keep the error checking

Reply:
I'm not sure this is cleaner because:
- we still need a nonlocal input_spec. Because we need to ferry this information from the .apply to the .forward
- adding additional outputs to forward means that we need to consider them in backward, which makes the backward function more complicated.

But I'm happy to try if you think this is better.

Reply:
I can buy that!